File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[438]找到字符串中所有字母异位词.scala
package leetcode.editor.cn
object FindAllAnagramsInAString {
//leetcode submit region begin(Prohibit modification and deletion)
  object Solution {
    import scala.collection.mutable
    def findAnagrams(s: String, p: String): List[Int] = {
      val need = p.groupBy(identity).view.mapValues(_.length).toMap
      val window = mutable.Map.empty[Char, Int]
      var left = 0
      var right = 0
      var valid = 0
      var result: List[Int] = Nil

      while (right < s.length) {
        val c = s(right)
        right += 1

        if (need.contains(c)) {
          window.updateWith(c) {
            case Some(v) =>
              Some(v + 1)
            case None => Some(1)
          }
          if (window(c) == need(c)) valid += 1
        }

        while (right - left >= p.length) {
          if (valid == need.size) {
            result = left :: result
          }

          val d = s(left)
          left += 1

          if (need.contains(d)) {
            window.updateWith(d) {
              case Some(v) =>
                if (v == need(d)) valid -= 1
                Some(v - 1)
              case None => None
            }
          }
        }
      }

      result
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}

File: cpp/leetcodecn/[438]找到字符串中所有字母异位词.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        unordered_map<char, int> need;
        for (auto c: p) need[c]++;

        unordered_map<char, int> window;
        vector<int> res;
        int left = 0;
        int right = 0;
        int valid = 0;

        while (right < s.size()) {
            auto ch = s[right];
            ++right;

            if (need.count(ch)) {
                window[ch]++;
                if (window[ch] == need[ch])
                    ++valid;
            }

            while (right - left >= p.size()) {
                if (valid == need.size()) res.push_back(left);

                auto d = s[left];
                ++left;

                if (need.contains(d)) {
                    if (window[d] == need[d]) --valid;
                    window[d]--;
                }
            }
        }

        return res;

    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}